flareon/src/utils/graph.rs

```diff
@@ -0,0 +1,134 @@
+use thiserror::Error;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, Error)]
+#[error("Cycle detected in the graph")]
+pub struct CycleDetected;
+
+pub fn apply_permutation<T>(items: &mut [T], order: &mut [usize]) {
+    for i in 0..order.len() {
+        let mut current = i;
+        let mut next = order[current];
+
+        while next != i {
+            // process the cycle
+            items.swap(current, next);
+            order[current] = current;
+
+            current = next;
+            next = order[current];
+        }
+
+        order[current] = current;
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub(crate) struct Graph {
```

Comment:
Since we have `petgraph` now, wouldn't it make sense to replace our `Graph` with theirs?